docs/blog/_posts/2016-12-05-implicit-function-types.md

```diff
@@ -0,0 +1,170 @@
+---
+layout: blog
+title: Implicit Function Types
+author: Martin Odersky
+authorImg: /images/martin.jpg
+---
+
+I just made the first pull request to add _implicit function types_ to
+Scala. I am pretty excited about it, because, citing the explanation
+of the pull request "This is the first step to bring comonadic
+abstraction to Scala". That's quite a mouthful, so I better explain what I
+mean by it.
+
+Let me try to explain the words in this sentence from right to left.
+
+*Scala*: I assume everyone who reads this understands that we mean the
+ programming language, not the opera house.
+
+*Abstraction*: The ability to name a concept and use just the name afterwards.
+
+*Comonadic*: In category theory, a _comonad_ is the dual of a
+_monad_. Roughly speaking, a monad is a way to wrap the result (or:
+outputs) of a computation in some other type. For instance
+`Future[T]` means that the result of type `T` will be produced at
+some later time on demand, or `Option[T]` indicates that the result
+might also be undefined.
+
+Dually, a _comonad_ allows to transform or
+enrich or otherwise manipulate the _inputs_ to a computation.
+The inputs are typically what a computation can access in its
+environment. Interesting tasks that are by nature comonadic are
+
+ - passing configuration data to the parts of a system that need them,
+ - managing capabilities for security critical tasks,
+ - wiring components up with dependency injection,
+ - defining the meanings of operations with type classes,
+ - more generally, passing any sort of context to a computation.
```

Comment:
I see a mention of my work on coeffects in the thread already, so I thought I'll add my perspective. I'm not all that familiar with Scala, so I'll stick to my Haskell/ML-style notation, but I hope that will be understandable.

**What are coeffects**
First of all, coeffects are, indeed, closely related to comonads (they are modelled using "indexed comonads", which is a generalisation of ordinary comonads). What coeffects add, is that they track more precisely what context is needed. The comonadic (or monadic) type `C a` or `M a` tells you that there is _some_ context-requirement or _some_ effect, but it does not tell you _what_ precisely. Coeffects add an annotation so that you have types such as `C {?p:int, ?q:int} a` where the annotation `{?p:int, ?q:int}` says you need two implicit parameters `?p` and `?q`. So you can see coeffects as more precise comonads.

**Are implicit parameters monads, comonads or coeffects**
Implicit parameters are one of the motivating examples in the work on coeffects, but they are just one (coeffects capture other interesting contextual properties). The interesting thing about implicit parameters is that they can *almost* be modelled by both Reader monad and Product comonad. With monads, you use functions `a -> M b` for some monad and with comonads, you use functions `C a -> b` for some comonad.

With Reader monad, we have `M a = State -> a` and so:

    a -> M b = a -> (State -> b)

With Product comonad, we have `C a = a * State` and so:

    C a -> b = a * State -> b

Now you can see that the two functions, `a * State -> b` and `a -> State -> b` are related (via currying)! 🎉 

There is one thing that reader monad does not let you do. It does not let you model the facts that implicit parameters (in GHC) support both lexical and dynamic scoping. Take for example this:

    let ?x = 1
    let f = (fun n ->?x + ?y)

If we model this using Reader monad and `?x` and `?y` mean "read implicit parameter from the monad", then the type of `f` will be a function that needs `?x` and `?y`. However, if we do this using comonads, the context available in the function body can combine context provided by the caller with the context provided by the declaration site - and so we can use `?x` from the declaration site and end up with a function that needs just `?y` (I'd be quite curious to see what Scala is doing here - are implicits just dynamically scoped or mix of both? 

**Links with more information**
There is a lot more about this than I can fit in a comment, so for those interested:

 * I wrote an [Interactive, reader-friendly online essay about coeffects](http://tomasp.net/coeffects/)
 * All the glorious details are in [my PhD thesis](https://github.com/coeffects/coeffects-thesis/blob/master/main.pdf) - relevant sections for the discussion above are 2.2, 3.2.1 and 5.6.
 * Someone already mentioned our [ICFP paper](http://tomasp.net/academic/papers/structural/), but the older [ICALP paper](http://tomasp.net/academic/papers/coeffects/) might be easier to read if you care only about implicit parameters.